disttask/framework/dispatcher/dispatcher.go

```diff
@@ -0,0 +1,412 @@
+
+const (
+	// DefaultConcurrency is the default concurrency for handling global task.
+	DefaultConcurrency = 8
+	// DefaultSubtaskConcurrency is the default concurrency for handling subtask.
+	DefaultSubtaskConcurrency = 16
+	// MaxSubtaskConcurrency is the maximum concurrency for handling subtask.
+	MaxSubtaskConcurrency     = 256
+	checkTaskFinishedInterval = 300 * time.Millisecond
+	checkTaskRunningInterval  = 300 * time.Millisecond
+)
+
+// Dispatch defines the interface for operations inside a dispatcher.
+type Dispatch interface {
+	// Start enables dispatching and monitoring mechanisms.
+	Start()
+	// Stop stops the dispatcher.
+	Stop()
+}
+
+func (d *dispatcher) getRunningGlobalTasks() map[int64]*proto.Task {
+	d.runningGlobalTasks.RLock()
+	defer d.runningGlobalTasks.RUnlock()
+	return d.runningGlobalTasks.tasks
+}
+
+func (d *dispatcher) isRunningGlobalTask(globalTaskID int64) bool {
+	d.runningGlobalTasks.RLock()
+	defer d.runningGlobalTasks.RUnlock()
+	_, ok := d.runningGlobalTasks.tasks[globalTaskID]
+	return ok
+}
+
+func (d *dispatcher) setRunningGlobalTasks(gTask *proto.Task) {
+	d.runningGlobalTasks.Lock()
+	defer d.runningGlobalTasks.Unlock()
+	d.runningGlobalTasks.tasks[gTask.ID] = gTask
+}
+
+func (d *dispatcher) delRunningGlobalTasks(globalTaskID int64) {
+	d.runningGlobalTasks.Lock()
+	defer d.runningGlobalTasks.Unlock()
+	delete(d.runningGlobalTasks.tasks, globalTaskID)
+}
+
+type dispatcher struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	gTaskMgr   *storage.GlobalTaskManager
+	subTaskMgr *storage.SubTaskManager
+	wg         tidbutil.WaitGroupWrapper
+
+	runningGlobalTasks struct {
+		sync.RWMutex
+		tasks map[int64]*proto.Task
+	}
+}
+
+// NewDispatcher creates a dispatcher struct.
+func NewDispatcher(ctx context.Context, globalTaskTable *storage.GlobalTaskManager, subtaskTable *storage.SubTaskManager) (*dispatcher, error) {
+	dispatcher := &dispatcher{
+		gTaskMgr:   globalTaskTable,
+		subTaskMgr: subtaskTable,
+	}
+	dispatcher.ctx, dispatcher.cancel = context.WithCancel(ctx)
+	dispatcher.runningGlobalTasks.tasks = make(map[int64]*proto.Task)
+
+	return dispatcher, nil
+}
+
+// Start implements Dispatch.Start interface.
+func (d *dispatcher) Start() {
+	d.wg.Run(func() {
+		d.DispatchTaskLoop()
+	})
+	d.wg.Run(func() {
+		d.DetectTaskLoop()
+	})
+}
+
+// Stop implements Dispatch.Stop interface.
+func (d *dispatcher) Stop() {
+	d.cancel()
+	d.wg.Wait()
+}
+
+// DispatchTaskLoop dispatches the global tasks.
+func (d *dispatcher) DispatchTaskLoop() {
+	logutil.BgLogger().Info("dispatch task loop start")
+	ticker := time.NewTicker(checkTaskRunningInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-d.ctx.Done():
+			logutil.BgLogger().Info("dispatch task loop exits", zap.Error(d.ctx.Err()))
+			return
+		case <-ticker.C:
+			cnt := len(d.getRunningGlobalTasks())
+
+			for cnt < DefaultConcurrency {
+				// TODO: Consider get all unfinished tasks.
+				gTask, err := d.gTaskMgr.GetNewTask()
```

Comment:
What if the dispatch crashes when a task's state changes to "running" ? `GetNewTask` only returns a task with state `Pending`

Reply:
This item does not have an interface on the storage side at present, and will be changed to get unfinished tasks later